python/ray/data/aggregate.py

```diff
@@ -92,42 +91,119 @@ def accumulate_block(a: AggType, block: Block) -> AggType:
         if finalize is None:
             finalize = lambda a: a  # noqa: E731
 
+        self.name = name
         self.init = init
         self.merge = merge
-        self.name = name
         self.accumulate_block = accumulate_block
         self.finalize = finalize
 
-    def _validate(self, schema: Optional[Union[type, "pa.lib.Schema"]]) -> None:
+    def _validate(self, schema: Optional["Schema"]) -> None:
         """Raise an error if this cannot be applied to the given schema."""
         pass
 
 
-class _AggregateOnKeyBase(AggregateFn):
-    def _set_key_fn(self, on: str):
-        self._key_fn = on
+@PublicAPI(stability="alpha")
+class AggregateFnV2(AggregateFn, Generic[AggType]):
+    """Provides an interface to implement efficient aggregations to be applied
+    to the dataset.
+
+    `AggregateFnV2` instances are passed to a Dataset's ``.aggregate(...)`` method to
+    perform aggregations by applying distributed aggregation algorithm:
+
+        - `aggregate_block` is applied to individual blocks, producing partial
+            aggregations.
+        - `combine` combines new partially aggregated value (previously returned
+            from `aggregate_block` partial aggregations into a singular partial
+            aggregation) with the previously stored accumulator.
+        - `finalize` transforms partial aggregation into its final state (for
+            some aggregations this is an identity transformation, ie no-op)
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        *,
+        on: Optional[str],
```

Comment:
this should be a list?

Reply:
Yes, all current impls are single-column though, so this extension is for the future